### PR TITLE
fixes issue #132 add evalulate_on_new_document

### PR DESCRIPTION
**Pull request by alexanderadam**
Hi @route,

so I thought that I might give [issue 132](https://github.com/rubycdp/ferrum/issues/132) a go.

As always: thank you for your work and [never give up](https://twitter.com/alexanderadam__/status/1362818462557622272) :wink:

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -124,6 +124,18 @@ browser.mouse
...
+Evaluate JavaScript to modify things before a page load:
+
+```ruby
+javascript = <<~JS
+  Object.defineProperty(navigator, 'languages', {
+    get: function() { return ['tlh']; }
+  });
+JS
+
+Browser.new(evaluate_on_new_document: javascript)
```
Comment (route) [inline README.md:136]: I think we already have `:extensions` we can say in the description of this newly added method that they are related, no need to introduce new option.

File: lib/ferrum/browser.rb [hunk 2]
```diff
@@ -77,6 +77,10 @@ def extensions
       end
     end
 
+    def evaluate_on_new_document
```
Comment (route) [inline lib/ferrum/browser.rb:80]: I think we should reuse what we have for extensions now and modify `@extensions` array calling this method

File: README.md (hunk 1 above)
Comment (alexanderadam) [reply README.md]: I'm not quite sure how you meant this exactly but I changed it how I understood it.
Can you check again please?

File: lib/ferrum/browser.rb [hunk 3]
```diff
@@ -77,6 +77,11 @@ def extensions
       end
     end
 
+    def evaluate_on_new_document(javascript = nil)
+      extensions << (block_given? ? yield : javascript)
```
Comment (route) [inline lib/ferrum/browser.rb:81]: block is not needed I guess, and we definitely should accept something so not need for `= nil`

File: lib/ferrum/browser.rb [hunk 4]
```diff
@@ -77,6 +77,11 @@ def extensions
...
+    end
+    alias evaluate_on_new_document= evaluate_on_new_document
```
Comment (route) [inline lib/ferrum/browser.rb:83]: `evaluate_on_new_document=` confuses me as I would think that I can override previously passed script, let's leave only `evaluate_on_new_document`

File: README.md [hunk 5]
```diff
@@ -124,6 +124,18 @@ browser.mouse
 browser.quit
 ```
 
+Evaluate JavaScript to modify things before a page load:
```
Comment (route) [inline README.md:127]: We have a section for JavaScript, let's move this method there

File: lib/ferrum/browser.rb (hunk 3 above)
Comment (alexanderadam) [reply lib/ferrum/browser.rb]: Oh really? I liked that syntax:

```ruby
browser.evaluate_on_new_document do
  <<~JS
    Object.defineProperty(navigator, 'languages', {
      get: function() { return ['tlh']; }
    });
  JS
end
```

I changed it accordingly though.

Comment (ttilberg) [reply lib/ferrum/browser.rb]: That syntax is almost still entirely valid, however by adding the block into the mix, there's a certain element of unnecessary indirection in the API. Questions that come to mind are "Does the block evaluate in a certain context? Are there args? How is this different from the `javascript` arg?"

You surely know this, but this is still valid:

```rb
browser.evaluate_on_new_document <<~JS
  Object.defineProperty(navigator, 'languages', {
    get: function() { return ['tlh']; }
  });
JS
```

And I leaves fewer questions when learning the API.

As a side note, I too love using the heredoc+highlight syntax for `js` strings. Particularly in Ferrum, it feels fantastic.

Comment (alexanderadam) [reply lib/ferrum/browser.rb]: > by adding the block into the mix, there's a certain element of unnecessary indirection in the API

For me blocks create a semantic boundary which goes further than "only" having a heredoc.
But again: I changed it anyway. :wink:

**Comment by route** [comment]
Looking good, thank you!

**Pull request merged**